dc/s2n-quic-dc/src/stream/server/udp.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{accept, InitialPacket};
+use crate::{
+    credentials::Credentials,
+    event::{self, EndpointPublisher as _, Subscriber},
+    msg,
+    packet::stream,
+    path::secret,
+    socket::recv::{descriptor, router::Router},
+    stream::{
+        endpoint,
+        environment::{udp, Environment},
+        recv::dispatch::{Allocator, Dispatch},
+        socket,
+    },
+};
+use s2n_quic_core::{
+    event::IntoEvent,
+    inet::{ExplicitCongestionNotification, SocketAddress},
+    time::Clock,
+    varint::VarInt,
+};
+use schnellru::LruMap;
+use std::{io, sync::Arc};
+use tracing::debug;
+
+pub struct Acceptor<Env, S, W>
+where
+    Env: Environment,
+    S: socket::application::Application,
+    W: socket::Socket,
+{
+    sender: accept::Sender<Env::Subscriber>,
+    env: Env,
+    secrets: secret::Map,
+    accept_flavor: accept::Flavor,
+    credential_cache: CredentialCache,
+    dispatch: Dispatch,
+    queues: Allocator,
+    is_open: bool,
+    packet: InitialPacket,
+    application_socket: Arc<S>,
+    worker_socket: Arc<W>,
+}
+
+impl<Env, S, W> Acceptor<Env, S, W>
+where
+    Env: Environment,
+    S: socket::application::Application,
+    W: socket::Socket,
+{
+    pub fn new(
+        env: Env,
+        sender: accept::Sender<Env::Subscriber>,
+        secrets: secret::Map,
+        accept_flavor: accept::Flavor,
+        queues: Allocator,
+        application_socket: Arc<S>,
+        worker_socket: Arc<W>,
+        credential_cache_size: u32,
+    ) -> Self {
+        let dispatch = queues.dispatcher();
+        let credential_cache = create_credential_cache(credential_cache_size);
+        let packet = InitialPacket::empty();
+        Self {
+            sender,
+            env,
+            secrets,
+            accept_flavor,
+            credential_cache,
+            dispatch,
+            queues,
+            is_open: true,
+            packet,
+            application_socket,
+            worker_socket,
+        }
+    }
+}
+
+impl<Env, S, W> Router for Acceptor<Env, S, W>
+where
+    Env: Environment,
+    Env::Subscriber: Clone,
+    S: socket::application::Application,
+    W: socket::Socket,
+{
+    #[inline]
+    fn is_open(&self) -> bool {
+        self.is_open
+    }
+
+    #[inline]
+    fn handle_stream_packet(
+        &mut self,
+        _remote_address: SocketAddress,
+        _ecn: ExplicitCongestionNotification,
+        packet: stream::decoder::Packet,
+    ) {
+        self.packet = packet.into();
+    }
+
+    #[inline]
+    fn dispatch_stream_packet(
+        &mut self,
+        _tag: stream::Tag,
+        _id: stream::Id,
+        credentials: Credentials,
+        segment: descriptor::Filled,
+    ) {
+        let credentials = CredentialsHashable(credentials);
+        if let Some(queue_id) = self.credential_cache.get(&credentials) {
```

Comment:
This doesn't really sound like a *credential* cache -- rather we're caching the queue ID a particular credential maps to. But even that seems off to me, because AFAICT we won't land on the same queue if the cache evicts the credential, we'll get mapped to some other newly allocated queue? So I'm not even really sure that this is a cache per-se. (I don't really understand how this works if we get evicted from the cache -- surely we need stream packets to go to a consistent queue?)

Reply:
Talked offline - i'm going to rework this as a secondary key mechanism in the queue allocator.